kubeflow/core/iap.libsonnet

```diff
@@ -143,15 +216,110 @@
       },
       data: {
-        "envoy-config.json": std.manifestJson($.parts(namespace).envoyConfig(audiences, disableJwt)),
+        "envoy-config.json": std.manifestJson($.parts(namespace).envoyConfig(disableJwt)),
+        // Script executed by init container to enable IAP. When finished, the configmap is patched with the JWT audience.
+        "iap-init.sh": |||
+          [ -z ${CLIENT_ID} ] && echo Error CLIENT_ID must be set && exit 1
+          [ -z ${CLIENT_SECRET} ] && echo Error CLIENT_SECRET must be set && exit 1
+          [ -z ${NAMESPACE} ] && echo Error NAMESPACE must be set && exit 1
+          [ -z ${SERVICE} ] && echo Error SERVICE must be set && exit 1
+
+          apk add --update jq
+          curl https://storage.googleapis.com/kubernetes-release/release/v1.9.4/bin/linux/amd64/kubectl > /usr/local/bin/kubectl && chmod +x /usr/local/bin/kubectl
+          
+          # Stagger init of replicas when acquiring lock
+          sleep $(( $RANDOM % 5 + 1 ))
+          
+          kubectl get svc ${SERVICE} -o json > service.json
+          LOCK=$(jq -r ".metadata.annotations.iaplock" service.json)
+          
+          NOW=$(date -u +'%s')
+          if [[ -z "${LOCK}" || "${LOCK}" == "null" ]]; then 
+            LOCK_T=$NOW
+          else
+            LOCK_T=$(echo "${LOCK}" | cut -d' ' -f2)
+          fi
+          LOCK_AGE=$(( $NOW - $LOCK_T ))
+          LOCK_TTL=120
+          if [[ -z "${LOCK}" || "${LOCK}" == "null" || "${LOCK_AGE}" -gt "${LOCK_TTL}" ]]; then
+            jq -r ".metadata.annotations.iaplock=\"$(hostname -s) ${NOW}\"" service.json > service_lock.json
+            kubectl apply -f service_lock.json 2>/dev/null
+            if [[ $? -eq 0 ]]; then
+              echo "Acquired lock on service annotation to update IAP."
+            else
+              echo "WARN: Failed to acquire lock on service annotation."
+              exit 1
+            fi
+          else
+            echo "WARN: Lock on service annotation already acquired by: $LOCK, age: $LOCK_AGE, TTL: $LOCK_TTL"
+            sleep 20
+            exit 1
+          fi
+          
+          PROJECT=$(curl -s -H "Metadata-Flavor: Google" http://metadata.google.internal/computeMetadata/v1/project/project-id)
+          if [ -z ${PROJECT} ]; then
+            echo Error unable to fetch PROJECT from compute metadata
+            exit 1
+          fi
+
+          PROJECT_NUM=$(curl -s -H "Metadata-Flavor: Google" http://metadata.google.internal/computeMetadata/v1/project/numeric-project-id)
+          if [ -z ${PROJECT_NUM} ]; then
+            echo Error unable to fetch PROJECT_NUM from compute metadata
+            exit 1
+          fi
+          
+          NODE_PORT=$(kubectl --namespace=${NAMESPACE} get svc ${SERVICE} -o jsonpath='{.spec.ports[0].nodePort}')
+          while [[ -z ${BACKEND_ID} ]];
+          do BACKEND_ID=$(gcloud compute --project=${PROJECT} backend-services list --filter=name~k8s-be-${NODE_PORT}- --format='value(id)');
```

Comment:
Does this use the Compute Engine default service account credentials? If so, can you add a comment in the iap.md document to mention that "Compute Engine default service account" should have Project Editor role or higher in IAM settings.

Reply:
It will use the service account for the GKE cluster, which by default is the Compute Engine default service account. I'll add a note about the permission required in the `iap.md`.